com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/antlr/ParserSemanticPredicatesUtil.java

```diff
@@ -0,0 +1,628 @@
+
+  /**
+   * Prevents instantiation of {@link AcfAntlrGeneratorFragmentUtil}.
+   */
+  private ParserSemanticPredicatesUtil() {
+    // Prevents creating an instance
+  }
+
+  /**
+   * Returns disambiguating semantic predicate if the given rule is a keyword rule.
+   *
+   * @param rule
+   *          Xtext grammar rule
+   * @return A string containing the semantic predicate or an empty string
+   */
+  public static String getPredicate(final AbstractRule rule) {
+    return getRulePredicate(rule, false);
+  }
+
+  /**
+   * Returns gated semantic predicate if the given rule is a keyword rule.
+   *
+   * @param ruleCall
+   *          Rule call grammar element
+   * @return A string containing the semantic predicate or an empty string
+   */
+  public static String getGatedPredicate(final AbstractElement ruleCall) {
+    return getRulePredicate(findPredicatedRule(ruleCall), true);
+  }
+
+  /**
+   * Returns a string with a gated semantic predicate for rules that replace a keyword or an alternative of keywords.
+   * Works for data type rules only.
+   * <p>
+   * <em>Note:</em> If the comment before the rule contains {@code  @KeywordRule(lookup) } then we generate a predicate in the rule so the rule looks like
+   *
+   * @param rule
+   *          Xtext grammar rule
+   * @param gated
+   *          Whether the inserted predicate should be a gated predicate
+   * @return A string containing the semantic predicate or an empty string {@code parserContext.getInput().LT(1).getText().equals(“lookup")}?=> } Multiple
+   *         keywords are also supported. Such as: {@code @KeywordRule(hide,show,show_per_bu) }
+   *         </p>
+   *         <p>
+   *         <em>Gated vs. Disambiguating predicates</em> On the first look a disambiguating semantic predicate would be a right choice. And indeed this is
+   *         true, however inserting them would be much more difficult. The reason is that we need not only insert them in the beginning of the current rule,
+   *         but also propagate before actions into alternatives that call this rule. We have to do the same for gated predicated, however there is one
+   *         important difference. If we have an alternative like {@code rule: (Keyword1 | Keyword2) Keyword3 ;} then we need to insert the predicate only
+   *         before the first alternative (Keyword1). If we insert disambiguating semantic predicate before both the exception is will throw will damage the
+   *         recovery for the entire rule i.e. will not try going to Keyword3. Clearly we need validating predicated in the beginning of each rule Keyword2,
+   *         Keyword3. This requires an extra analysis which might be a next step. Gated predicates we can safely insert before Keyword1 and Keyword2. This will
+   *         have no negative impact. We still need validating predicates in Keyword rules themselves.
+   *         </p>
+   *         <p>
+   *         </p>

```

Comment:
can we remove this 'paragrah' tag which has no text?

Reply:
ok